src/envalid.ts

```diff
@@ -0,0 +1,146 @@
+//   validators: { [K in keyof T]: ValidatorSpec<T[K]> },
+//   options: StrictCleanOptions,
+// ): Readonly<T> & CleanEnv
+function cleanEnv<T>(
+  env: unknown,
+  specs: { [K in keyof T]: ValidatorSpec<T[K]> },
+  options: CleanOptions = { middleware: defaultMiddlewares },
+): Readonly<T> & CleanEnv {
+  let output: any = {}
+  let defaultNodeEnv = ''
+  const errors: any = {}
+  const varKeys = Object.keys(specs) as Array<keyof T>
+  const rawNodeEnv = (env as any).NODE_ENV
+
+  // FIXME: make this opt-in, as an exported util?
+  // If validation for NODE_ENV isn't specified, use the default validation:
+  // @ts-expect-error
+  if (!varKeys.includes('NODE_ENV')) {
+    defaultNodeEnv = validateVar({
+      name: 'NODE_ENV',
+      spec: str({ choices: ['development', 'test', 'production'] }),
+      rawValue: rawNodeEnv || 'production',
+    })
+  }
+
+  for (const k of varKeys) {
+    const spec = specs[k]
+    const usingDevDefault = rawNodeEnv !== 'production' && spec.hasOwnProperty('devDefault')
+    const devDefault = usingDevDefault ? spec.devDefault : undefined
+    // @ts-ignore FIXME
+    const rawValue = env[k] ?? (devDefault === undefined ? spec.default : devDefault)
+
+    // Default values can be anything falsy (including an explicitly set undefined), without
+    // triggering validation errors:
+    const usingFalsyDefault =
+      (spec.hasOwnProperty('default') && spec.default === rawValue) ||
+      (usingDevDefault && devDefault === rawValue)
+
+    try {
+      // @ts-ignore FIXME
+      if (rawValue === testOnlySymbol) {
+        throw new EnvMissingError(formatSpecDescription(spec))
+      }
+
+      if (rawValue === undefined) {
+        if (!usingFalsyDefault) {
+          throw new EnvMissingError(formatSpecDescription(spec))
+        }
+
+        output[k] = undefined
+      } else {
+        // @ts-ignore FIXME
+        output[k] = validateVar({ name: k, spec, rawValue })
+      }
+    } catch (err) {
+      if (options.reporter === null) throw err
+      errors[k] = err
+    }
+  }
+
+  // TODO: replace all the below with middleware: cleanedEnv => modifiedEnv
+
+  // Provide is{Prod/Dev/Test} properties for more readable NODE_ENV checks
+  // Node that isDev and isProd are just aliases to isDevelopment and isProduction
+  const computedNodeEnv = defaultNodeEnv || output.NODE_ENV
+  Object.defineProperties(output, {
+    isDevelopment: { value: computedNodeEnv === 'development' },
+    isDev: { value: computedNodeEnv === 'development' },
+    isProduction: { value: computedNodeEnv === 'production' },
+    isProd: { value: computedNodeEnv === 'production' },
+    isTest: { value: computedNodeEnv === 'test' },
+  })
+
+  const reporter = options.reporter || defaultReporter
+  reporter({ errors, env: output })
+
+  // Apply middlewares to transform the validated env object
+  if (!options.middleware?.length) return output
+  return options.middleware.reduce((acc, mw) => mw(acc, env), output)
```

Comment:
I think using middlewares like this will destroy the integrity of the return type?

Reply:
Also if you have built-in defaults, like the freeze and proxy, it gets a bit tricky when you override the middlewares since you always then also have to pass in the built-in middlewares which one "always" wants

Reply:
If you go the middleware-approach, maybe error reporter etc should be a middleware as well

Reply:
I think there's probably a good argument to applying the freeze after middleware runs, so it's never omitted. For the proxy, I think it's possible people might want to opt out of it.

In terms of overriding the default middleware, I'm exporting `defaultMiddlewares` so it can simply be extended before being passed in the cleanEnv. So something like this would work:

```
const middleware = [myCustomMW1, myCustomMW2, ...defaultMiddlewares]
```